atomic_reactor/plugins/post_group_manifests.py

```diff
@@ -318,62 +152,32 @@ def sort_annotations(self):
-                    continue
-
-                platforms = sorted_digests.setdefault(hostname, {})
-                repos = platforms.setdefault(plat, [])
-                repos.append(digest)
-
-        sources = {}
-        for registry in self.registries:
-            hostname = registry_hostname(registry)
-            platforms = sorted_digests.get(hostname, {})
-
-            if set(platforms) != all_platforms:
-                raise RuntimeError("Missing platforms for registry {}: found {}, expected {}"
-                                   .format(registry, sorted(platforms), sorted(all_platforms)))
-
-            selected_digests = {}
-            for p, repos in platforms.items():
-                selected_digests[p] = sorted(repos, key=lambda d: d['repository'])[0]
-
-            sources[registry] = selected_digests
-
-        return sources
-
-    def get_registry_session(self, registry):
-        registry_conf = self.registries[registry]
-
-        insecure = registry_conf.get('insecure', False)
-        secret_path = registry_conf.get('secret')
-
-        return RegistrySession(registry, insecure=insecure,
-                               dockercfg_path=secret_path,
-                               access=('pull', 'push'))
+        return self.manifest_util.sort_annotations(all_annotations)
 
     def run(self):
-        digests = dict()
+        list_json = ""
+        # No other plugin really needs the list of repos, but we should return it for
+        # unit tests
+        repos = []
+
         for registry, source in self.sort_annotations().items():
-            session = self.get_registry_session(registry)
+            session = self.manifest_util.get_registry_session(registry)
 
             if self.group:
-                repo, digest = self.group_manifests_and_tag(session, source)
+                repo, digest, list_json = self.group_manifests_and_tag(session, source)
                 self.log.debug("repo: %s digest: %s", repo, digest)
-                digests[repo] = digest
+                repos.append(repo)
             else:
                 found = False
                 if len(source) != 1:
                     raise RuntimeError('Without grouping only one source is expected')
                 for digest in source.values():
-                    self.tag_manifest_into_registry(session, digest)
+                    source_digest = digest['digest']
+                    source_repo = digest['repository']
+                    self.manifest_util.tag_manifest_into_registry(session, source_digest,
+                                                                  source_repo,
+                                                                  self.workflow.tag_conf.images)
                     found = True
                 if not found:
                     raise ValueError('Failed to find any platform')
-        return digests
+        return {'digest': digest, 'repos': repos, 'manifest_list': list_json}
```

Comment:
Sorry, just spotted that `digest` here is a dict, I'd thought it was a string (it gets reused with different types). Can we arrange for `digest` to always be a string, so that the 'digest' in our return dict is always a string?

Reply:
it's actually a ManifestDigest object.  returning a manifest digest was kind of the point of group_manifests, or so I thought.

is it okay to return it as `{'manifest_digest': digest, 'manifest_list': list_json}` ?

Reply:
Oh! I had missed that it's a ManifestDigest object.

Yes, that looks fine to return.